File: src/path_planning/example/discrete_trajectory_decomposition.py
import sys, os, math
sys.path.append(os.path.join(os.path.dirname(__file__), ".."))

from user_interface.matrixIO import read_data_from_txt
from user_interface.plotUtils import set_ax_equal

import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.art3d import Poly3DCollection


def axis_angle(axis, angle):
    axis = np.array(axis[0:4])
    axis = axis / np.linalg.norm(axis)

    kx = np.zeros((3,3))
    kx[0][1], kx[0][2] = -axis[2], axis[1]
    kx[1][0], kx[1][2] = axis[2], -axis[0]
    kx[2][0], kx[2][1] = -axis[1], axis[0]

    eye = np.eye(3)
    rot = eye + np.sin(angle)*kx + (1-np.cos(angle))*kx@kx

    return rot

def plot_circle(ax, p1, p2, p3):
    p1, p2, p3 = [np.array(p1), np.array(p2), np.array(p3)]

    a = p1 - p2
    b = p3 - p2
    # 半径过大时，直接绘制直线
    if (np.cross(a,b).dot(np.cross(a,b)) < 1e-10):
        ax.plot([p1[0], p3[0]], [p1[1], p3[1]], [p1[2], p3[2]])
        return
    r = np.cross(a.dot(a)*b - b.dot(b)*a, np.cross(a,b)) / (2*np.cross(a,b).dot(np.cross(a,b)))
    #  print(math.sqrt(r.dot(r)))

    center = r + p2
    #  print("center = ", center)
    op1, op2, op3 = p1-center, p2-center, p3-center

    # 法向
    norm = np.cross(op1, op3)
    norm = norm / np.linalg.norm(norm)

    # O_p1 与 O_p3 的夹角
    theta = np.arccos(op1.dot(op3) / np.linalg.norm(op1) / np.linalg.norm(op3))
    # p2 是否在 <p1_O_p3 的劣弧内
    #  if (np.cross(op1, op2).dot(np.cross(op2,op3)) < 0):
    if (op1.dot(op2) <=0 or op2.dot(op3) <= 0):
        theta = 2*np.pi - theta
        norm *= -1
    #  print("norm = ", norm)
    #  print(theta * 180 / np.pi)

    #  ax.scatter([p1[0],p2[0],p3[0]], [p1[1],p2[1],p3[1]], [p1[2],p2[2],p3[2]])

    q = np.linspace(0, theta, 50)
    x, y, z = np.array([]), np.array([]), np.array([])
    for i in q:
        rot = axis_angle(norm, i)
        pnt = rot@op1 + center
        x, y, z = np.append(x, pnt[0]), np.append(y, pnt[1]), np.append(z, pnt[2])
    ax.plot(x, y, z)


def plot_mbs(ax, dataFile="final_22.txt"):
    # 读取 cpp 生成的测试数据，在项目根目录下运行`script/..`
    mat = read_data_from_txt(dataPath+dataFile).transpose()
    idx = read_data_from_txt(savePath+"sep_idx")
    apIdx = read_data_from_txt(savePath+"arc_point_idx")

    ax.set_xlabel('X')
    ax.set_ylabel('Y')
    ax.set_zlabel('Z')

    # 原始数据图
    ax.plot(mat[0,:], mat[1,:], mat[2,:], linestyle = ':')

    # 分割点
    begIdx = int(idx[0][0])
    for i in range(len(idx[0])):
        endIdx = int(i+1) % len(idx[0])
        x = [mat[0, int(idx[0][i])], mat[0, int(idx[0][endIdx])]]
        y = [mat[1, int(idx[0][i])], mat[1, int(idx[0][endIdx])]]
        z = [mat[2, int(idx[0][i])], mat[2, int(idx[0][endIdx])]]
        #  ax.plot(x, y, z, c='r')
        if (len(idx[0]) < 15):
            ax.scatter(mat[0, int(idx[0][i])],mat[1, int(idx[0][i])],mat[2, int(idx[0][i])])

    #  for i in [0]:
    for i in range(len(idx[0])):
        begIdx = int(idx[0][i])
        if (i == len(idx[0])-1):
            break
        midIdx = int(apIdx[0][i])
        endIdx = int(idx[0][i+1])
        plot_circle(ax, mat[:, begIdx], mat[:, midIdx], mat[:, endIdx])

        begIdx = endIdx

    set_ax_equal(ax)


def plot_mpc(ax):
    pnt = read_data_from_txt(savePath+"mpc").transpose()
    idx = read_data_from_txt(savePath+"mpc_dp_idx")
    ax.plot(pnt[0,:], pnt[1,:])
    ax.scatter(pnt[0,:], pnt[1,:])

    for i in range(len(idx[0])):
        if (i < len(idx[0])-1):
            x = [pnt[0, int(idx[0][i])], pnt[0, int(idx[0][i+1])]]
            y = [pnt[1, int(idx[0][i])], pnt[1, int(idx[0][i+1])]]
            ax.plot(x, y)
        ax.scatter(pnt[0, int(idx[0][i])],pnt[1, int(idx[0][i])], c='r')

def plot_trans(ax):
    pnt = read_data_from_txt(savePath+"trans_pnt").transpose()
    ax.plot(*pnt)

    set_ax_equal(ax)
    return

def plot_discrete_traj(ax):
    pnt = read_data_from_txt(savePath+"node_pnt").transpose()
    arc = read_data_from_txt(savePath+"arc_info").transpose()
    ax.scatter(*pnt)
    for i in range(len(arc[0])):
        if arc[3,i] > 0:
            center = arc[0:3,i]
            op1 = pnt[:,i] - center
            theta = np.linalg.norm(arc[4:,i])
            norm = arc[4:,i] / theta
            q = np.linspace(0, theta, 50)
            x, y, z = np.array([]), np.array([]), np.array([])
            for i in q:
                rot = axis_angle(norm, i)
                pnt = rot@op1 + center
                x, y, z = np.append(x, pnt[0]), np.append(y, pnt[1]), np.append(z, pnt[2])
            ax.plot(x, y, z, c = '#1f77b4', linestyle = ':')
        else:
            pkg = [[pnt[j,i], pnt[j,i+1]] for j in range(3)]
            ax.plot(*pkg, c = '#1f77b4', linestyle = ':')

    set_ax_equal(ax)
    return

if __name__ == "__main__":
    dataPath = "data/curve_decomposition/"
    savePath = "build/data/curve_decomposition/"

    dataFile = sys.argv[1] if (len(sys.argv) > 1) else "final_22.txt"

    fig = plt.figure('空间三角形',figsize=(8,6))
    ax_mbs =plt.subplot(211,projection='3d')
    ax_mpc =plt.subplot(212)

    #  ax_mbs =plt.subplot(211,projection='3d')
    #  ax_trans =plt.subplot(111,projection='3d')
    #  plot_discrete_traj(ax_mbs)
    #  plot_trans(ax_mbs)

    plot_mbs(ax_mbs, dataFile)
    plot_mpc(ax_mpc)


    plt.show()

File: src/path_planning/example/discrete_trajectory_decomposition.cpp
#include "user_interface/user_interface.h"
#include "path_planning/curve_decomposition.h"

// 配置参数
// 直线近似误差
const double linearErr = 2;
// 曲率分割误差
const double curveErr = 2e-1;
// 数据存放路径
const std::string dataPath = "data/curve_decomposition/";
// 文件保存路径
const std::string savePath = "build/data/curve_decomposition/";

// 读取数据的文件名称
std::string dataFile;
// 数据点
Eigen::MatrixXd pntMat;
// 数据点
std::vector<Eigen::Vector3d> fitPnt;
// 离散轨迹类
DiscreteTrajectory traj;

int discrete_trajectory_decomposition(int argc, char** argv);
void discrete_trajectory_transition();
// 结果可视化
void decomposition_visualization();
// 贝塞尔曲线过渡
void bezier_transition(std::list<Eigen::Vector3d>& nodePnt, std::list<Eigen::Vector<double,7>>& arcInfo, double smooth);

int main(int argc, char** argv) {
  // discrete_trajectory_decomposition(argc, argv);
  // discrete_trajectory_transition();

  // Eigen::Vector3d op1 = (beg - cen).normalized(), op2 = (mid - cen).normalized(), op3 = (end - cen).normalized();
  // std::cout << "op1 = " << op1.transpose() << std::endl;
  // std::cout << "op3 = " << op3.transpose() << std::endl;
  // std::cout << "norm = " << (op1.cross(op3)).transpose() << std::endl;
  // std::cout << (85.68) * M_PI / 180 << std::endl;

  return 0;
}

void discrete_trajectory_transition() {
  std::list<Eigen::Vector3d> nodePnt;
  std::list<Eigen::Vector<double,7>> arcInfo;

  nodePnt.push_back({0, 0, 0});
  nodePnt.push_back({1, 0, 0});
  nodePnt.push_back({1, 0, 1});
  arcInfo.push_back({0, 0, 0, -1, 0, 0, 0});
  arcInfo.push_back({1, 0, 0.5, 0.5, 0, M_PI, 0});

  std::ofstream nodePntFile(savePath + "node_pnt", std::ios::trunc);
  for (const auto& node : nodePnt) {
    nodePntFile << node.transpose() << std::endl;
  }
  bezier_transition(nodePnt, arcInfo, 0.2);

  std::ofstream arcInfoFile(savePath + "arc_info", std::ios::trunc);
  for (const auto& arc : arcInfo) {
    arcInfoFile << arc.transpose() << std::endl;
  }

  std::cout << "After ================> " << std::endl;
  std::ofstream transPntFile(savePath + "trans_pnt", std::ios::trunc);
  for (auto& pnt : nodePnt) {
    transPntFile << pnt.transpose() << std::endl;
  }
  return;
}

int discrete_trajectory_decomposition(int argc, char** argv) {
  dataFile = argc > 1 ? argv[1] : "final_22.txt";
  std::cout << "Read data from: " << dataFile << std::endl;

  if (int flag = read_eigen_from_file(dataPath + dataFile, pntMat); flag > 0) {
    std::cout << "Read data failed with code: " << flag << std::endl;
    return 1;
  }
  pntMat.transposeInPlace();

  // 以向量格式保存拟合点
  fitPnt = std::vector<Eigen::Vector3d>(pntMat.cols());
  for (int i=0; i<pntMat.cols(); ++i) {
    fitPnt[i] = pntMat.col(i);
  }
  traj.pntList = fitPnt;

  decomposition_visualization();

  traj.perform_trajectory_decomposition(linearErr, curveErr);
  traj.calc_trajectory_pose({0,0,1});

  std::list<Eigen::Vector3d> nodePnt;
  std::list<Eigen::Vector<double,7>> arcInfo;
  traj.export_trajectory(nodePnt, arcInfo);

  for (auto& pnt : nodePnt) {
    std::cout << pnt.transpose() << std::endl;
  }

  bezier_transition(nodePnt, arcInfo, 20);

  std::cout << "After ================> " << std::endl;
  std::ofstream transPntFile(savePath + "trans_pnt", std::ios::trunc);
  for (auto& pnt : nodePnt) {
    transPntFile << pnt.transpose() << std::endl;
  }

  return 0;
}

void decomposition_visualization() {
  // Blurred segment splition for midpoitn curve -> dominant points
  std::vector<size_t> dpIdx = maximal_blurred_segment_split(fitPnt, linearErr);
  // Get midpoint curve of the maximal blurred segment
  std::vector<Eigen::Vector3d> mpc =  convert_into_tangent_space(fitPnt, dpIdx);
  // Blurred segment splition for midpoitn curve
  std::vector<size_t> idxList = maximal_blurred_segment_split(mpc, curveErr);
  // 圆弧和直线段分隔点的索引
  std::vector<size_t> sepIdx = std::vector<size_t>(idxList.size(), dpIdx.back());
  for (size_t i=0; i<idxList.size()-1; ++i) {
    sepIdx[i] = dpIdx[idxList[i]];
  }
  // 逐段进行圆弧近似
  std::vector<std::pair<size_t, double>> pair = arc_approximation(fitPnt, sepIdx);

  // Mid point curve: 累计弦长 - 累计夹角
  std::ofstream mpcFile(savePath + "mpc", std::ios::trunc);
  for (size_t i=0; i<mpc.size(); ++i) {
    mpcFile << mpc[i][0] << " " << mpc[i][1] << std::endl;
  }
  // MPC 曲线中 DP 点的索引
  std::ofstream mpcDpFile(savePath + "mpc_dp_idx", std::ios::trunc);
  for (size_t i=0; i<idxList.size(); ++i) {
    mpcDpFile << idxList[i] << std::endl;
  }
  // 分割点索引
  std::ofstream sepIdxFile(savePath + "sep_idx", std::ios::trunc);
  for (size_t i=0; i<sepIdx.size(); ++i) {
    sepIdxFile << sepIdx[i] << std::endl;
  }
  // 圆弧点索引
  std::ofstream apIdxFile(savePath + "arc_point_idx");
  for (auto& p : pair) {
    apIdxFile << p.first << std::endl;
  }
}

void bezier_transition(std::list<Eigen::Vector3d>& nodePnt, std::list<Eigen::Vector<double,7>>& arcInfo, double smooth) {
  // 轨迹节点迭代器
  std::list<Eigen::Vector3d>::iterator endIte = nodePnt.begin();
  std::list<Eigen::Vector3d>::iterator begIte = endIte++;
  // 轨迹信息迭代器
  std::list<Eigen::Vector<double,7>>::iterator infoIte = arcInfo.begin();

  // 上一段轨迹终点, 下一段轨迹的起点处的过渡点
  Eigen::Vector3d preMark = *begIte, aftMark(0,0,0);
  // 过渡点处的切线方向
  Eigen::Vector3d preNorm(0,0,0), aftNorm(0,0,0);
  // 贝塞尔曲线过渡段长度
  double tranLen = 0.0;

  // 记录第一段轨迹点终点信息
  Eigen::Vector3d begPnt = *begIte, endPnt = *endIte;
  Eigen::Vector<double,7> info = *infoIte;
  if (info(3) > 1000 || info(3) < 0) {
    preNorm = (endPnt - begPnt).normalized();
    preMark = endPnt - preNorm*smooth;
  } else {
    // 切向量, 圆心
    Eigen::Vector3d normal = info.tail(3).normalized(), center = info.head(3);
    // 半径
    double radius = (center - begPnt).norm();
    // 过渡圆弧对应的圆心角
    double ang = smooth / radius;
    // 过渡点位置
    preMark = Eigen::AngleAxisd(-ang, normal) * (endPnt - center) + center;
    // 切线方向
    preNorm = normal.cross((preMark - center).normalized());
  }
  tranLen = smooth * preNorm.dot((endPnt - preMark).normalized());

  // 遍历每一段轨迹
  infoIte++;
  for (auto curInfoIte = infoIte; curInfoIte != arcInfo.end(); ++curInfoIte) {
    // 更新迭代器
    begIte = endIte++;
    // infoIte++;
    // 更新轨迹信息
    begPnt = *begIte, endPnt = *endIte;
    info = *curInfoIte;

    // 计算起点处的过渡点位置和切向量
    if (info(3) > 1000 || info(3) < 0) {
      // 起点处的过渡点
      aftNorm = (endPnt - begPnt).normalized();
      // 过渡点处的切向量
      aftMark = begPnt + aftNorm*smooth;
    } else {
      // 切向量, 圆心
      Eigen::Vector3d normal = info.tail(3).normalized(), center = info.head(3);
      // 半径
      double radius = (center - begPnt).norm();
      // 过渡圆弧对应的圆心角
      double ang = smooth / radius;
      // 过渡点位置
      aftMark = Eigen::AngleAxisd(ang, normal) * (begPnt - center) + center;
      // 切线方向
      aftNorm = normal.cross((aftMark - center).normalized());
    }
    tranLen = std::min(tranLen, smooth * aftNorm.dot((aftMark - begPnt).normalized()));

    double vel = 0.01, freq = 1;
    // 计算过渡曲线
    BezierCurve bezier;
    bezier.construct_from_vct(preMark, preNorm, aftMark, aftNorm, 0.3*tranLen);
    std::vector<double> paraVec;
    bezier.discrete_arc_length(paraVec, 0, 1, 1e-2);
    std::vector<double> cumuChordLength;
    double curveLen = bezier.get_chord_length(paraVec, cumuChordLength);
    size_t numSegm = std::ceil(curveLen / vel * freq);
    std::vector<double> samplePara = bezier.get_uniform_sample(paraVec, cumuChordLength, curveLen / numSegm);

    // 修改上一段终点
    *begIte = bezier.get_point(samplePara[0]);
    // *begIte = preMark;
    // nodePnt.insert(endIte, aftMark);
    // arcInfo.insert(curInfoIte, {0,0,0,-1,0,0,0});
    // 修改旋转角度
    // 插入过渡线段
    for (size_t i=1; i<samplePara.size(); ++i) {
      nodePnt.insert(endIte, bezier.get_point(samplePara[i]));
      arcInfo.insert(infoIte, {0,0,0,-1,0,0,0});
    }

    // 更新终点处的过渡点和切线方向
    if (info(3) > 1000 || info(3) < 0) {
      preNorm = (endPnt - begPnt).normalized();
      preMark = endPnt - preNorm*smooth;
    } else {
      // 切向量, 圆心
      Eigen::Vector3d normal = info.tail(3).normalized(), center = info.head(3);
      // 半径
      double radius = (center - begPnt).norm();
      // 过渡圆弧对应的圆心角
      double ang = smooth / radius;
      // 过渡点位置
      preMark = Eigen::AngleAxisd(-ang, normal) * (endPnt - center) + center;
      // 切线方向
      preNorm = normal.cross((preMark - center).normalized());
    }
    tranLen = smooth * preNorm.dot((endPnt - preMark).normalized());
  }
}

